{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: cache_decorator in /home/odococo/anaconda3/lib/python3.7/site-packages (1.2.5)\n",
      "Requirement already satisfied: compress-pickle in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache_decorator) (1.1.1)\n",
      "Requirement already satisfied: openpyxl in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache_decorator) (3.0.3)\n",
      "Requirement already satisfied: xlrd in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache_decorator) (1.2.0)\n",
      "Requirement already satisfied: pandas in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache_decorator) (1.0.3)\n",
      "Requirement already satisfied: dict-hash in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache_decorator) (1.0.8)\n",
      "Requirement already satisfied: compress-json in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache_decorator) (1.0.3)\n",
      "Requirement already satisfied: numpy in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache_decorator) (1.18.2)\n",
      "Requirement already satisfied: jdcal in /home/odococo/anaconda3/lib/python3.7/site-packages (from openpyxl->cache_decorator) (1.4.1)\n",
      "Requirement already satisfied: et-xmlfile in /home/odococo/anaconda3/lib/python3.7/site-packages (from openpyxl->cache_decorator) (1.0.1)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->cache_decorator) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->cache_decorator) (2019.3)\n",
      "Requirement already satisfied: six>=1.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from python-dateutil>=2.6.1->pandas->cache_decorator) (1.14.0)\n",
      "Requirement already satisfied: encodeproject in /home/odococo/anaconda3/lib/python3.7/site-packages (1.0.17)\n",
      "Requirement already satisfied: tqdm in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject) (4.44.1)\n",
      "Requirement already satisfied: requests in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject) (2.23.0)\n",
      "Requirement already satisfied: humanize in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject) (2.3.0)\n",
      "Requirement already satisfied: cache-decorator in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject) (1.2.5)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject) (2020.4.5.1)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject) (1.25.8)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject) (3.0.4)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject) (2.9)\n",
      "Requirement already satisfied: compress-json in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject) (1.0.3)\n",
      "Requirement already satisfied: pandas in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject) (1.0.3)\n",
      "Requirement already satisfied: dict-hash in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject) (1.0.8)\n",
      "Requirement already satisfied: compress-pickle in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject) (1.1.1)\n",
      "Requirement already satisfied: numpy in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject) (1.18.2)\n",
      "Requirement already satisfied: xlrd in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject) (1.2.0)\n",
      "Requirement already satisfied: openpyxl in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject) (3.0.3)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->cache-decorator->encodeproject) (2019.3)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->cache-decorator->encodeproject) (2.8.1)\n",
      "Requirement already satisfied: jdcal in /home/odococo/anaconda3/lib/python3.7/site-packages (from openpyxl->cache-decorator->encodeproject) (1.4.1)\n",
      "Requirement already satisfied: et-xmlfile in /home/odococo/anaconda3/lib/python3.7/site-packages (from openpyxl->cache-decorator->encodeproject) (1.0.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from python-dateutil>=2.6.1->pandas->cache-decorator->encodeproject) (1.14.0)\n",
      "Requirement already satisfied: crr_labels in /home/odococo/anaconda3/lib/python3.7/site-packages (1.0.8)\n",
      "Requirement already satisfied: encodeproject in /home/odococo/anaconda3/lib/python3.7/site-packages (from crr_labels) (1.0.17)\n",
      "Requirement already satisfied: pandas in /home/odococo/anaconda3/lib/python3.7/site-packages (from crr_labels) (1.0.3)\n",
      "Requirement already satisfied: cache-decorator in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject->crr_labels) (1.2.5)\n",
      "Requirement already satisfied: tqdm in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject->crr_labels) (4.44.1)\n",
      "Requirement already satisfied: humanize in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject->crr_labels) (2.3.0)\n",
      "Requirement already satisfied: requests in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject->crr_labels) (2.23.0)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->crr_labels) (1.18.2)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->crr_labels) (2019.3)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->crr_labels) (2.8.1)\n",
      "Requirement already satisfied: xlrd in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject->crr_labels) (1.2.0)\n",
      "Requirement already satisfied: openpyxl in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject->crr_labels) (3.0.3)\n",
      "Requirement already satisfied: dict-hash in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject->crr_labels) (1.0.8)\n",
      "Requirement already satisfied: compress-json in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject->crr_labels) (1.0.3)\n",
      "Requirement already satisfied: compress-pickle in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject->crr_labels) (1.1.1)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject->crr_labels) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject->crr_labels) (1.25.8)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject->crr_labels) (2020.4.5.1)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject->crr_labels) (2.9)\n",
      "Requirement already satisfied: six>=1.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from python-dateutil>=2.6.1->pandas->crr_labels) (1.14.0)\n",
      "Requirement already satisfied: et-xmlfile in /home/odococo/anaconda3/lib/python3.7/site-packages (from openpyxl->cache-decorator->encodeproject->crr_labels) (1.0.1)\n",
      "Requirement already satisfied: jdcal in /home/odococo/anaconda3/lib/python3.7/site-packages (from openpyxl->cache-decorator->encodeproject->crr_labels) (1.4.1)\n",
      "Requirement already satisfied: pybwtool in /home/odococo/anaconda3/lib/python3.7/site-packages (1.0.8)\n",
      "Requirement already satisfied: numpy in /home/odococo/anaconda3/lib/python3.7/site-packages (from pybwtool) (1.18.2)\n",
      "Requirement already satisfied: pandas in /home/odococo/anaconda3/lib/python3.7/site-packages (from pybwtool) (1.0.3)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->pybwtool) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->pybwtool) (2019.3)\n",
      "Requirement already satisfied: six>=1.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from python-dateutil>=2.6.1->pandas->pybwtool) (1.14.0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: epigenomic_dataset in /home/odococo/anaconda3/lib/python3.7/site-packages (1.1.4)\n",
      "Requirement already satisfied: pybwtool>=1.0.8 in /home/odococo/anaconda3/lib/python3.7/site-packages (from epigenomic_dataset) (1.0.8)\n",
      "Requirement already satisfied: encodeproject>=1.0.9 in /home/odococo/anaconda3/lib/python3.7/site-packages (from epigenomic_dataset) (1.0.17)\n",
      "Requirement already satisfied: crr-labels>=1.0.7 in /home/odococo/anaconda3/lib/python3.7/site-packages (from epigenomic_dataset) (1.0.8)\n",
      "Requirement already satisfied: notipy-me>=1.3.11 in /home/odococo/anaconda3/lib/python3.7/site-packages (from epigenomic_dataset) (1.3.14)\n",
      "Requirement already satisfied: numpy in /home/odococo/anaconda3/lib/python3.7/site-packages (from pybwtool>=1.0.8->epigenomic_dataset) (1.18.2)\n",
      "Requirement already satisfied: pandas in /home/odococo/anaconda3/lib/python3.7/site-packages (from pybwtool>=1.0.8->epigenomic_dataset) (1.0.3)\n",
      "Requirement already satisfied: cache-decorator in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject>=1.0.9->epigenomic_dataset) (1.2.5)\n",
      "Requirement already satisfied: requests in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject>=1.0.9->epigenomic_dataset) (2.23.0)\n",
      "Requirement already satisfied: humanize in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject>=1.0.9->epigenomic_dataset) (2.3.0)\n",
      "Requirement already satisfied: tqdm in /home/odococo/anaconda3/lib/python3.7/site-packages (from encodeproject>=1.0.9->epigenomic_dataset) (4.44.1)\n",
      "Requirement already satisfied: environments-utils in /home/odococo/anaconda3/lib/python3.7/site-packages (from notipy-me>=1.3.11->epigenomic_dataset) (1.0.2)\n",
      "Requirement already satisfied: tabulate in /home/odococo/anaconda3/lib/python3.7/site-packages (from notipy-me>=1.3.11->epigenomic_dataset) (0.8.7)\n",
      "Requirement already satisfied: userinput in /home/odococo/anaconda3/lib/python3.7/site-packages (from notipy-me>=1.3.11->epigenomic_dataset) (1.0.13)\n",
      "Requirement already satisfied: validate-email in /home/odococo/anaconda3/lib/python3.7/site-packages (from notipy-me>=1.3.11->epigenomic_dataset) (1.3)\n",
      "Requirement already satisfied: validators in /home/odococo/anaconda3/lib/python3.7/site-packages (from notipy-me>=1.3.11->epigenomic_dataset) (0.14.3)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->pybwtool>=1.0.8->epigenomic_dataset) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->pybwtool>=1.0.8->epigenomic_dataset) (2019.3)\n",
      "Requirement already satisfied: dict-hash in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject>=1.0.9->epigenomic_dataset) (1.0.8)\n",
      "Requirement already satisfied: xlrd in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject>=1.0.9->epigenomic_dataset) (1.2.0)\n",
      "Requirement already satisfied: openpyxl in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject>=1.0.9->epigenomic_dataset) (3.0.3)\n",
      "Requirement already satisfied: compress-json in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject>=1.0.9->epigenomic_dataset) (1.0.3)\n",
      "Requirement already satisfied: compress-pickle in /home/odococo/anaconda3/lib/python3.7/site-packages (from cache-decorator->encodeproject>=1.0.9->epigenomic_dataset) (1.1.1)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject>=1.0.9->epigenomic_dataset) (1.25.8)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject>=1.0.9->epigenomic_dataset) (2.9)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject>=1.0.9->epigenomic_dataset) (3.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->encodeproject>=1.0.9->epigenomic_dataset) (2020.4.5.1)\n",
      "Requirement already satisfied: validate-version-code in /home/odococo/anaconda3/lib/python3.7/site-packages (from userinput->notipy-me>=1.3.11->epigenomic_dataset) (1.0.4)\n",
      "Requirement already satisfied: ipython in /home/odococo/anaconda3/lib/python3.7/site-packages (from userinput->notipy-me>=1.3.11->epigenomic_dataset) (7.13.0)\n",
      "Requirement already satisfied: IPy in /home/odococo/anaconda3/lib/python3.7/site-packages (from userinput->notipy-me>=1.3.11->epigenomic_dataset) (1.0)\n",
      "Requirement already satisfied: six>=1.4.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from validators->notipy-me>=1.3.11->epigenomic_dataset) (1.14.0)\n",
      "Requirement already satisfied: decorator>=3.4.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from validators->notipy-me>=1.3.11->epigenomic_dataset) (4.4.2)\n",
      "Requirement already satisfied: et-xmlfile in /home/odococo/anaconda3/lib/python3.7/site-packages (from openpyxl->cache-decorator->encodeproject>=1.0.9->epigenomic_dataset) (1.0.1)\n",
      "Requirement already satisfied: jdcal in /home/odococo/anaconda3/lib/python3.7/site-packages (from openpyxl->cache-decorator->encodeproject>=1.0.9->epigenomic_dataset) (1.4.1)\n",
      "Requirement already satisfied: pickleshare in /home/odococo/anaconda3/lib/python3.7/site-packages (from ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (0.7.5)\n",
      "Requirement already satisfied: traitlets>=4.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (4.3.3)\n",
      "Requirement already satisfied: pygments in /home/odococo/anaconda3/lib/python3.7/site-packages (from ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (2.6.1)\n",
      "Requirement already satisfied: setuptools>=18.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (46.1.3.post20200330)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (3.0.4)\n",
      "Requirement already satisfied: pexpect; sys_platform != \"win32\" in /home/odococo/anaconda3/lib/python3.7/site-packages (from ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (4.8.0)\n",
      "Requirement already satisfied: jedi>=0.10 in /home/odococo/anaconda3/lib/python3.7/site-packages (from ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (0.16.0)\n",
      "Requirement already satisfied: backcall in /home/odococo/anaconda3/lib/python3.7/site-packages (from ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (0.1.0)\n",
      "Requirement already satisfied: ipython-genutils in /home/odococo/anaconda3/lib/python3.7/site-packages (from traitlets>=4.2->ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (0.2.0)\n",
      "Requirement already satisfied: wcwidth in /home/odococo/anaconda3/lib/python3.7/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (0.1.9)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pexpect; sys_platform != \"win32\"->ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (0.6.0)\n",
      "Requirement already satisfied: parso>=0.5.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from jedi>=0.10->ipython->userinput->notipy-me>=1.3.11->epigenomic_dataset) (0.6.2)\n",
      "Requirement already satisfied: ucsc_genomes_downloader in /home/odococo/anaconda3/lib/python3.7/site-packages (1.1.18)\n",
      "Requirement already satisfied: tqdm in /home/odococo/anaconda3/lib/python3.7/site-packages (from ucsc_genomes_downloader) (4.44.1)\n",
      "Requirement already satisfied: pandas in /home/odococo/anaconda3/lib/python3.7/site-packages (from ucsc_genomes_downloader) (1.0.3)\n",
      "Requirement already satisfied: dateparser in /home/odococo/anaconda3/lib/python3.7/site-packages (from ucsc_genomes_downloader) (0.7.4)\n",
      "Requirement already satisfied: requests in /home/odococo/anaconda3/lib/python3.7/site-packages (from ucsc_genomes_downloader) (2.23.0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->ucsc_genomes_downloader) (2.8.1)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->ucsc_genomes_downloader) (1.18.2)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->ucsc_genomes_downloader) (2019.3)\n",
      "Requirement already satisfied: regex!=2019.02.19 in /home/odococo/anaconda3/lib/python3.7/site-packages (from dateparser->ucsc_genomes_downloader) (2020.4.4)\n",
      "Requirement already satisfied: tzlocal in /home/odococo/anaconda3/lib/python3.7/site-packages (from dateparser->ucsc_genomes_downloader) (2.0.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->ucsc_genomes_downloader) (2020.4.5.1)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->ucsc_genomes_downloader) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->ucsc_genomes_downloader) (1.25.8)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->ucsc_genomes_downloader) (2.9)\n",
      "Requirement already satisfied: six>=1.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from python-dateutil>=2.6.1->pandas->ucsc_genomes_downloader) (1.14.0)\n",
      "Requirement already satisfied: keras_bed_sequence in /home/odococo/anaconda3/lib/python3.7/site-packages (1.1.1)\n",
      "Requirement already satisfied: keras-mixed-sequence>=1.0.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from keras_bed_sequence) (1.0.6)\n",
      "Requirement already satisfied: pandas in /home/odococo/anaconda3/lib/python3.7/site-packages (from keras_bed_sequence) (1.0.3)\n",
      "Requirement already satisfied: numba in /home/odococo/anaconda3/lib/python3.7/site-packages (from keras_bed_sequence) (0.48.0)\n",
      "Requirement already satisfied: numpy in /home/odococo/anaconda3/lib/python3.7/site-packages (from keras_bed_sequence) (1.18.2)\n",
      "Requirement already satisfied: tensorflow in /home/odococo/anaconda3/lib/python3.7/site-packages (from keras_bed_sequence) (2.1.0)\n",
      "Requirement already satisfied: ucsc-genomes-downloader>=1.1.13 in /home/odococo/anaconda3/lib/python3.7/site-packages (from keras_bed_sequence) (1.1.18)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->keras_bed_sequence) (2019.3)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->keras_bed_sequence) (2.8.1)\n",
      "Requirement already satisfied: setuptools in /home/odococo/anaconda3/lib/python3.7/site-packages (from numba->keras_bed_sequence) (46.1.3.post20200330)\n",
      "Requirement already satisfied: llvmlite<0.32.0,>=0.31.0dev0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from numba->keras_bed_sequence) (0.31.0)\n",
      "Requirement already satisfied: tensorboard<2.2.0,>=2.1.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (2.1.1)\n",
      "Requirement already satisfied: wheel>=0.26; python_version >= \"3\" in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (0.34.2)\n",
      "Requirement already satisfied: keras-applications>=1.0.8 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (1.0.8)\n",
      "Requirement already satisfied: google-pasta>=0.1.6 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (0.2.0)\n",
      "Requirement already satisfied: gast==0.2.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (0.2.2)\n",
      "Requirement already satisfied: opt-einsum>=2.3.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (3.2.0)\n",
      "Requirement already satisfied: keras-preprocessing>=1.1.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (1.1.0)\n",
      "Requirement already satisfied: astor>=0.6.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (0.8.1)\n",
      "Requirement already satisfied: termcolor>=1.1.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (1.1.0)\n",
      "Requirement already satisfied: protobuf>=3.8.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (3.11.4)\n",
      "Requirement already satisfied: scipy==1.4.1; python_version >= \"3\" in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (1.4.1)\n",
      "Requirement already satisfied: tensorflow-estimator<2.2.0,>=2.1.0rc0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (2.1.0)\n",
      "Requirement already satisfied: absl-py>=0.7.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (0.9.0)\n",
      "Requirement already satisfied: six>=1.12.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (1.14.0)\n",
      "Requirement already satisfied: grpcio>=1.8.6 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (1.28.1)\n",
      "Requirement already satisfied: wrapt>=1.11.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorflow->keras_bed_sequence) (1.12.1)\n",
      "Requirement already satisfied: requests in /home/odococo/anaconda3/lib/python3.7/site-packages (from ucsc-genomes-downloader>=1.1.13->keras_bed_sequence) (2.23.0)\n",
      "Requirement already satisfied: tqdm in /home/odococo/anaconda3/lib/python3.7/site-packages (from ucsc-genomes-downloader>=1.1.13->keras_bed_sequence) (4.44.1)\n",
      "Requirement already satisfied: dateparser in /home/odococo/anaconda3/lib/python3.7/site-packages (from ucsc-genomes-downloader>=1.1.13->keras_bed_sequence) (0.7.4)\n",
      "Requirement already satisfied: google-auth<2,>=1.6.3 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorboard<2.2.0,>=2.1.0->tensorflow->keras_bed_sequence) (1.13.1)\n",
      "Requirement already satisfied: werkzeug>=0.11.15 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorboard<2.2.0,>=2.1.0->tensorflow->keras_bed_sequence) (1.0.1)\n",
      "Requirement already satisfied: google-auth-oauthlib<0.5,>=0.4.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorboard<2.2.0,>=2.1.0->tensorflow->keras_bed_sequence) (0.4.1)\n",
      "Requirement already satisfied: markdown>=2.6.8 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tensorboard<2.2.0,>=2.1.0->tensorflow->keras_bed_sequence) (3.2.1)\n",
      "Requirement already satisfied: h5py in /home/odococo/anaconda3/lib/python3.7/site-packages (from keras-applications>=1.0.8->tensorflow->keras_bed_sequence) (2.10.0)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->ucsc-genomes-downloader>=1.1.13->keras_bed_sequence) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->ucsc-genomes-downloader>=1.1.13->keras_bed_sequence) (1.25.8)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->ucsc-genomes-downloader>=1.1.13->keras_bed_sequence) (2.9)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests->ucsc-genomes-downloader>=1.1.13->keras_bed_sequence) (2020.4.5.1)\n",
      "Requirement already satisfied: regex!=2019.02.19 in /home/odococo/anaconda3/lib/python3.7/site-packages (from dateparser->ucsc-genomes-downloader>=1.1.13->keras_bed_sequence) (2020.4.4)\n",
      "Requirement already satisfied: tzlocal in /home/odococo/anaconda3/lib/python3.7/site-packages (from dateparser->ucsc-genomes-downloader>=1.1.13->keras_bed_sequence) (2.0.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from google-auth<2,>=1.6.3->tensorboard<2.2.0,>=2.1.0->tensorflow->keras_bed_sequence) (0.2.8)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from google-auth<2,>=1.6.3->tensorboard<2.2.0,>=2.1.0->tensorflow->keras_bed_sequence) (4.0.0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: rsa<4.1,>=3.1.4 in /home/odococo/anaconda3/lib/python3.7/site-packages (from google-auth<2,>=1.6.3->tensorboard<2.2.0,>=2.1.0->tensorflow->keras_bed_sequence) (4.0)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard<2.2.0,>=2.1.0->tensorflow->keras_bed_sequence) (1.3.0)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pyasn1-modules>=0.2.1->google-auth<2,>=1.6.3->tensorboard<2.2.0,>=2.1.0->tensorflow->keras_bed_sequence) (0.4.8)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard<2.2.0,>=2.1.0->tensorflow->keras_bed_sequence) (3.1.0)\n",
      "Requirement already satisfied: keras_tqdm in /home/odococo/anaconda3/lib/python3.7/site-packages (2.0.1)\n",
      "Requirement already satisfied: Keras in /home/odococo/anaconda3/lib/python3.7/site-packages (from keras_tqdm) (2.3.1)\n",
      "Requirement already satisfied: tqdm in /home/odococo/anaconda3/lib/python3.7/site-packages (from keras_tqdm) (4.44.1)\n",
      "Requirement already satisfied: keras-applications>=1.0.6 in /home/odococo/anaconda3/lib/python3.7/site-packages (from Keras->keras_tqdm) (1.0.8)\n",
      "Requirement already satisfied: numpy>=1.9.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from Keras->keras_tqdm) (1.18.2)\n",
      "Requirement already satisfied: keras-preprocessing>=1.0.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from Keras->keras_tqdm) (1.1.0)\n",
      "Requirement already satisfied: h5py in /home/odococo/anaconda3/lib/python3.7/site-packages (from Keras->keras_tqdm) (2.10.0)\n",
      "Requirement already satisfied: pyyaml in /home/odococo/anaconda3/lib/python3.7/site-packages (from Keras->keras_tqdm) (5.3.1)\n",
      "Requirement already satisfied: scipy>=0.14 in /home/odococo/anaconda3/lib/python3.7/site-packages (from Keras->keras_tqdm) (1.4.1)\n",
      "Requirement already satisfied: six>=1.9.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from Keras->keras_tqdm) (1.14.0)\n",
      "Requirement already satisfied: minepy in /home/odococo/anaconda3/lib/python3.7/site-packages (1.2.4)\n",
      "Requirement already satisfied: numpy>=1.3.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from minepy) (1.18.2)\n",
      "Requirement already satisfied: boruta in /home/odococo/anaconda3/lib/python3.7/site-packages (0.3)\n",
      "Requirement already satisfied: scikit-learn>=0.17.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from boruta) (0.22.2.post1)\n",
      "Requirement already satisfied: numpy>=1.10.4 in /home/odococo/anaconda3/lib/python3.7/site-packages (from boruta) (1.18.2)\n",
      "Requirement already satisfied: scipy>=0.17.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from boruta) (1.4.1)\n",
      "Requirement already satisfied: joblib>=0.11 in /home/odococo/anaconda3/lib/python3.7/site-packages (from scikit-learn>=0.17.1->boruta) (0.14.1)\n",
      "Requirement already satisfied: prince in /home/odococo/anaconda3/lib/python3.7/site-packages (0.7.0)\n",
      "Requirement already satisfied: matplotlib>=3.0.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from prince) (3.2.1)\n",
      "Requirement already satisfied: scipy>=1.3.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from prince) (1.4.1)\n",
      "Requirement already satisfied: numpy>=1.17.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from prince) (1.18.2)\n",
      "Requirement already satisfied: pandas>=1.0.3 in /home/odococo/anaconda3/lib/python3.7/site-packages (from prince) (1.0.3)\n",
      "Requirement already satisfied: scikit-learn>=0.22.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from prince) (0.22.2.post1)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from matplotlib>=3.0.2->prince) (2.4.7)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from matplotlib>=3.0.2->prince) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/odococo/anaconda3/lib/python3.7/site-packages (from matplotlib>=3.0.2->prince) (0.10.0)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from matplotlib>=3.0.2->prince) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas>=1.0.3->prince) (2019.3)\n",
      "Requirement already satisfied: joblib>=0.11 in /home/odococo/anaconda3/lib/python3.7/site-packages (from scikit-learn>=0.22.1->prince) (0.14.1)\n",
      "Requirement already satisfied: six in /home/odococo/anaconda3/lib/python3.7/site-packages (from cycler>=0.10->matplotlib>=3.0.2->prince) (1.14.0)\n",
      "Requirement already satisfied: MulticoreTSNE in /home/odococo/anaconda3/lib/python3.7/site-packages (0.1)\n",
      "Requirement already satisfied: cffi in /home/odococo/anaconda3/lib/python3.7/site-packages (from MulticoreTSNE) (1.14.0)\n",
      "Requirement already satisfied: numpy in /home/odococo/anaconda3/lib/python3.7/site-packages (from MulticoreTSNE) (1.18.2)\n",
      "Requirement already satisfied: pycparser in /home/odococo/anaconda3/lib/python3.7/site-packages (from cffi->MulticoreTSNE) (2.20)\n",
      "Requirement already satisfied: tsnecuda in /home/odococo/anaconda3/lib/python3.7/site-packages (0.1.1)\n",
      "Requirement already satisfied: numpy>=1.14.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from tsnecuda) (1.18.2)\n",
      "Requirement already satisfied: barplots in /home/odococo/anaconda3/lib/python3.7/site-packages (1.1.4)\n",
      "Requirement already satisfied: tqdm in /home/odococo/anaconda3/lib/python3.7/site-packages (from barplots) (4.44.1)\n",
      "Requirement already satisfied: sanitize-ml-labels in /home/odococo/anaconda3/lib/python3.7/site-packages (from barplots) (1.0.9)\n",
      "Requirement already satisfied: humanize in /home/odococo/anaconda3/lib/python3.7/site-packages (from barplots) (2.3.0)\n",
      "Requirement already satisfied: scipy in /home/odococo/anaconda3/lib/python3.7/site-packages (from barplots) (1.4.1)\n",
      "Requirement already satisfied: pillow in /home/odococo/anaconda3/lib/python3.7/site-packages (from barplots) (7.0.0)\n",
      "Requirement already satisfied: matplotlib in /home/odococo/anaconda3/lib/python3.7/site-packages (from barplots) (3.2.1)\n",
      "Requirement already satisfied: pandas in /home/odococo/anaconda3/lib/python3.7/site-packages (from barplots) (1.0.3)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /home/odococo/anaconda3/lib/python3.7/site-packages (from scipy->barplots) (1.18.2)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from matplotlib->barplots) (2.4.7)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/odococo/anaconda3/lib/python3.7/site-packages (from matplotlib->barplots) (0.10.0)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from matplotlib->barplots) (2.8.1)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/odococo/anaconda3/lib/python3.7/site-packages (from matplotlib->barplots) (1.2.0)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from pandas->barplots) (2019.3)\n",
      "Requirement already satisfied: six in /home/odococo/anaconda3/lib/python3.7/site-packages (from cycler>=0.10->matplotlib->barplots) (1.14.0)\n",
      "Requirement already satisfied: ray in /home/odococo/anaconda3/lib/python3.7/site-packages (0.8.4)\n",
      "Requirement already satisfied: redis>=3.3.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (3.4.1)\n",
      "Requirement already satisfied: aiohttp in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (3.6.2)\n",
      "Requirement already satisfied: colorama in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (0.4.3)\n",
      "Requirement already satisfied: numpy>=1.16 in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (1.18.2)\n",
      "Requirement already satisfied: py-spy>=0.2.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (0.3.3)\n",
      "Requirement already satisfied: google in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (2.0.3)\n",
      "Requirement already satisfied: filelock in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (3.0.12)\n",
      "Requirement already satisfied: pyyaml in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (5.3.1)\n",
      "Requirement already satisfied: jsonschema in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (3.2.0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: protobuf>=3.8.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (3.11.4)\n",
      "Requirement already satisfied: click in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (7.1.1)\n",
      "Requirement already satisfied: grpcio in /home/odococo/anaconda3/lib/python3.7/site-packages (from ray) (1.28.1)\n",
      "Requirement already satisfied: multidict<5.0,>=4.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from aiohttp->ray) (4.7.5)\n",
      "Requirement already satisfied: async-timeout<4.0,>=3.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from aiohttp->ray) (3.0.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from aiohttp->ray) (19.3.0)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from aiohttp->ray) (1.4.2)\n",
      "Requirement already satisfied: chardet<4.0,>=2.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from aiohttp->ray) (3.0.4)\n",
      "Requirement already satisfied: beautifulsoup4 in /home/odococo/anaconda3/lib/python3.7/site-packages (from google->ray) (4.8.2)\n",
      "Requirement already satisfied: six>=1.11.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from jsonschema->ray) (1.14.0)\n",
      "Requirement already satisfied: importlib-metadata; python_version < \"3.8\" in /home/odococo/anaconda3/lib/python3.7/site-packages (from jsonschema->ray) (1.5.0)\n",
      "Requirement already satisfied: pyrsistent>=0.14.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from jsonschema->ray) (0.16.0)\n",
      "Requirement already satisfied: setuptools in /home/odococo/anaconda3/lib/python3.7/site-packages (from jsonschema->ray) (46.1.3.post20200330)\n",
      "Requirement already satisfied: idna>=2.0 in /home/odococo/anaconda3/lib/python3.7/site-packages (from yarl<2.0,>=1.0->aiohttp->ray) (2.9)\n",
      "Requirement already satisfied: soupsieve>=1.2 in /home/odococo/anaconda3/lib/python3.7/site-packages (from beautifulsoup4->google->ray) (2.0)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/odococo/anaconda3/lib/python3.7/site-packages (from importlib-metadata; python_version < \"3.8\"->jsonschema->ray) (2.2.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install cache_decorator\n",
    "!pip install encodeproject\n",
    "!pip install crr_labels\n",
    "!pip install pybwtool\n",
    "!pip install epigenomic_dataset\n",
    "!pip install ucsc_genomes_downloader\n",
    "!pip install keras_bed_sequence\n",
    "!pip install keras_tqdm\n",
    "!pip install minepy\n",
    "!pip install boruta\n",
    "!pip install prince\n",
    "!pip install MulticoreTSNE\n",
    "!pip install tsnecuda\n",
    "!pip install barplots\n",
    "!pip install ray\n",
    "!pip install tensorflow>=2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data retrieval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from epigenomic_dataset import load_epigenomes\n",
    "from ucsc_genomes_downloader import Genome\n",
    "from keras_bed_sequence import BedSequence\n",
    "from sklearn.impute import KNNImputer\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics.pairwise import euclidean_distances\n",
    "from sklearn.decomposition import PCA\n",
    "from scipy.stats import pearsonr\n",
    "from scipy.stats import spearmanr\n",
    "from scipy.stats import entropy\n",
    "from minepy import MINE\n",
    "from boruta import BorutaPy\n",
    "from prince import MFA\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from typing import Dict, List, Tuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_data(cell_line: str, dataset: str = \"fantom\", window_size: int = 200) -> Dict[str, pd.DataFrame]:\n",
    "    regions = [\"promoters\", \"enhancers\"]\n",
    "    epigenomes = {}\n",
    "    labels = {}\n",
    "    for region in regions:\n",
    "        epigenome, label = load_epigenomes(\n",
    "            cell_line=cell_line,\n",
    "            dataset=dataset,\n",
    "            regions=region,\n",
    "            window_size=window_size\n",
    "        )\n",
    "        epigenomes.update({region: epigenome})\n",
    "        labels.update({region: label})\n",
    "    return epigenomes, labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_bed(data: pd.DataFrame)-> pd.DataFrame:\n",
    "    \"\"\"Return bed coordinates from given dataset.\"\"\"\n",
    "    return data.reset_index()[data.index.names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genome(assembly: str = \"hg19\") -> Genome:\n",
    "    return Genome(assembly)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_hot_encode(data: pd.DataFrame, genome: Genome, nucleotides : str = \"actg\", window_size: int = 200)-> np.ndarray:\n",
    "    \"\"\"Set to one only the nucletoide, zero the others\"\"\"\n",
    "    return np.array(BedSequence(\n",
    "        genome,\n",
    "        bed=to_bed(data),\n",
    "        nucleotides=nucleotides,\n",
    "        batch_size=1\n",
    "    )).reshape(-1, window_size * len(nucleotides)).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_dataframe(data: np.ndarray, window_size : int = 200, nucleotides : str = \"actg\")-> pd.DataFrame:\n",
    "    return pd.DataFrame(\n",
    "        data,\n",
    "        columns = [\n",
    "            f\"{i}{nucleotide}\"\n",
    "            for i in range(window_size)\n",
    "            for nucleotide in nucleotides\n",
    "        ]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences(epigenomes: np.ndarray, genome: Genome, window_size: int = 200) -> Dict[str, pd.DataFrame]:\n",
    "    {\n",
    "    region: to_dataframe(\n",
    "        one_hot_encode(data, genome, window_size),\n",
    "        window_size\n",
    "    )\n",
    "    for region, data in epigenomes.items()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "info = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_retrieval(cell_line: str = 'K562') -> Dict[str, pd.DataFrame]:\n",
    "    global info\n",
    "    print('Downloading data...')\n",
    "    epigenomes, labels = info.get('data') or download_data(cell_line)\n",
    "    info['data'] = (epigenomes, labels)\n",
    "    print('Downloading genome...')\n",
    "    genome = info.get('genome') or get_genome()\n",
    "    info['genome'] = genome\n",
    "    print('Getting dataframe...')\n",
    "    sequences = info.get('sequences') or get_sequences(epigenomes, genome)\n",
    "    info['sequences'] = sequences\n",
    "    print('Finished!')\n",
    "    return epigenomes, labels, sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading data...\n",
      "Downloading genome...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Loading chromosomes for genome hg19', layout=Layout(flex=â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r",
      "Getting dataframe...\n"
     ]
    }
   ],
   "source": [
    "data_retrieval()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data elaboration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def overfitting_risk(epigenomes: Dict[str, pd.DataFrame], threshold: int = 1) -> bool:\n",
    "    valid = False\n",
    "    for region, data in epigenomes.items():\n",
    "        rate = data[0] / data[1]\n",
    "        print(f'Rate for {region} is {rate}')\n",
    "        valid = valid or (rate < threshold)\n",
    "    return valid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def nan_check(epigenomes: Dict[str, pd.DataFrame], threshold: int = 10) -> None:\n",
    "    for region, x in epigenomes.items():\n",
    "        print(\"\\n\".join((\n",
    "            f\"Nan values report for {region} data:\",\n",
    "            f\"In the document there are {x.isna().values.sum()} NaN values out of {x.values.size} values.\",\n",
    "            f\"The sample (row) with most values has {x.isna().sum(axis=0).max()} NaN values out of {x.shape[1]} values.\",\n",
    "            f\"The feature (column) with most values has {x.isna().sum().max()} NaN values out of {x.shape[0]} values.\"\n",
    "        )))\n",
    "        print(\"=\"*80)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fit_constant(data: pd.DataFrame, value: int) -> pd.DataFrame:\n",
    "    return df.fillna(value)\n",
    "\n",
    "\n",
    "def fit_media(data: pd.DataFrame) -> pd.DataFrame:\n",
    "    return fit_constant(data, data.mean())\n",
    "\n",
    "\n",
    "def fit_median(data: pd.DataFrame) -> pd.DataFrame:\n",
    "    return fit_constant(data, data.median())\n",
    "\n",
    "\n",
    "def fit_mode(data: pd.DataFrame) -> pd.DataFrame:\n",
    "    return fit_constant(data, data.mode())\n",
    "\n",
    "\n",
    "def fit_neighbours(data: pd.DataFrame, neighbours: int = 5) -> pd.DataFrame:\n",
    "    return pd.DataFrame(KNNImputer(n_neighbours=neighbours).fit_transform(data.values), \n",
    "                        columns=data.columns,\n",
    "                        index=data.index\n",
    "                       )\n",
    "\n",
    "\n",
    "def fit_missing(epigenomes: Dict[str, pd.DataFrame]) -> pd.DataFrame:\n",
    "    for region, data in epigenomes.items():\n",
    "        epigenomes[region] = fit_neighbours(data)\n",
    "    return epigenomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_class_balance(labels: Dict[str, pd.DataFrame]) -> None:\n",
    "    fig, axes = plt.subplots(ncols=2, figsize=(10, 5))\n",
    "\n",
    "    for axis, (region, y) in zip(axes.ravel(), labels.items()):\n",
    "        y.hist(ax=axis, bins=3)\n",
    "        axis.set_title(f\"Classes count in {region}\")\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drop_constant_features(epigenomes: Dict[str, pd.DataFrame]) -> Dict[str, pd.DataFrame]:\n",
    "    def drop(df:pd.Dataframe) -> pd.DataFrame:\n",
    "        return df.loc[:, (df != df.iloc[0]).any()]\n",
    "    for region, data in epigenomes.items():\n",
    "        result = drop_constant_features(data)\n",
    "        if data.shape[1] != result.shape[1]:\n",
    "            print(f\"Features in {region} were constant and had to be dropped!\")\n",
    "            epigenomes[region] = result\n",
    "        else:\n",
    "            print(f\"No constant features were found in {region}!\")\n",
    "    return epigenomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_z_scoring(epigenomes: Dict[str, pd.DataFrame]) -> Dict[str, pd.DataFrame]:\n",
    "    def robust_zscoring(df: pd.DataFrame) -> pd.DataFrame:\n",
    "        return pd.DataFrame(\n",
    "            RobustScaler().fit_transform(df.values),\n",
    "            columns=df.columns,\n",
    "            index=df.index\n",
    "        )\n",
    "    return {region: robust_zscoring(data) for region, data in epigenomes.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drop_uncorrelated(epigenomes: Dict[str, pd.DataFrame], p_value_threshold: float = 0.01, correlation_threshold: float = 0.05) -> Dict[str, pd.DataFrame]:\n",
    "    uncorrelated = {region: set() for region in epigenomes}\n",
    "    \n",
    "    def pearson():\n",
    "        for region, data in epigenomes.items():\n",
    "            for column in tqdm(data.columns, desc=f\"Running Pearson test for {region}\", dynamic_ncols=True, leave=False):\n",
    "                correlation, p_value = pearsonr(data[column].values.ravel(), labels[region].values.ravel())\n",
    "                if p_value > p_value_threshold:\n",
    "                    print(region, column, correlation)\n",
    "                    uncorrelated[region].add(column)\n",
    "    \n",
    "    def spearman():\n",
    "        for region, data in epigenomes.items():\n",
    "            for column in tqdm(data.columns, desc=f\"Running Spearman test for {region}\", dynamic_ncols=True, leave=False):\n",
    "                correlation, p_value = spearmanr(data[column].values.ravel(), labels[region].values.ravel())\n",
    "                if p_value > p_value_threshold:\n",
    "                    print(region, column, correlation)\n",
    "                uncorrelated[region].add(column)\n",
    "    \n",
    "    def mine():\n",
    "        for column in tqdm(uncorrelated[region], desc=f\"Running MINE test for {region}\", dynamic_ncols=True, leave=False):\n",
    "            mine = MINE()\n",
    "            mine.compute_score(x[column].values.ravel(), labels[region].values.ravel())\n",
    "            score = mine.mic()\n",
    "            if score < correlation_threshold:\n",
    "                print(region, column, score)\n",
    "            else:\n",
    "                uncorrelated[region].remove(column)\n",
    "                \n",
    "    def drop():\n",
    "        for region, data in epigenomes.items():\n",
    "            epigenomes[region] = data.drop(columns=[col for col in uncorrelated[region] if col in x.columns])\n",
    "        return epigenomes\n",
    "                \n",
    "    pearson()\n",
    "    spearman()\n",
    "    mine()\n",
    "    return drop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drop_too_correlated(epigenomes: Dict[str, pd.DataFrame], p_value_threshold: float = 0.01, correlation_threshold: float = 0.95) -> Dict[str, pd.DataFrame]:\n",
    "    extremely_correlated = {region: set() for region in epigenomes}\n",
    "    scores = {region: [] for region in epigenomes}\n",
    "    \n",
    "    def pearson():\n",
    "        for region, x in epigenomes.items():\n",
    "            for i, column in tqdm(\n",
    "                    enumerate(x.columns),\n",
    "                    total=len(x.columns), \n",
    "                    desc=f\"Running Pearson test for {region}\", \n",
    "                    dynamic_ncols=True, \n",
    "                    leave=False):\n",
    "                for feature in x.columns[i+1:]:\n",
    "                    correlation, p_value = pearsonr(x[column].values.ravel(), x[feature].values.ravel())\n",
    "                    correlation = np.abs(correlation)\n",
    "                    scores[region].append((correlation, column, feature))\n",
    "                    if p_value < p_value_threshold and correlation > correlation_threshold:\n",
    "                        print(region, column, feature, correlation)\n",
    "                        if entropy(x[column]) > entropy(x[feature]):\n",
    "                            extremely_correlated[region].add(feature)\n",
    "                        else:\n",
    "                            extremely_correlated[region].add(column)\n",
    "                        \n",
    "    return {region: sorted(score, key=lambda x: np.abs(x[0]), reverse=True) for region, score in scores.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def show(epigenomes: Dict[str, pd.DataFrame], scores: Dict[str, List[Tuple]]):\n",
    "    def get_data(region:str, how_many: int, from_start: bool = True) -> List[str]:\n",
    "        data = scores[region][:how_many] if from_start else scores[region][-how_many:]\n",
    "        data = list(zip(data))[1:]\n",
    "        columns = list(set([column for row in data for column in row]))\n",
    "        return columns\n",
    "    \n",
    "    def plot(how_many: int, correlated: bool) -> None:\n",
    "        for region, data in epigenomes.items():\n",
    "            print(f\"Most {'correlated' if correlated else 'uncorrelated'} features from {region} epigenomes\")\n",
    "            sns.pairplot(pd.concat([\n",
    "                x[get_data(region, how_many, correlated)],\n",
    "                labels[region],\n",
    "            ], axis=1), hue=labels[region].columns[0])\n",
    "            plt.show()\n",
    "    \n",
    "    def correlated(how_many: int):\n",
    "        plot(how_many, True)\n",
    "        \n",
    "    def uncorrelated(how_many: int):\n",
    "        plot(how_many, False)\n",
    "        \n",
    "    def get_top_most_different(dist, n: int):\n",
    "        return np.argsort(-np.mean(dist, axis=1).flatten())[:n]\n",
    "    \n",
    "    def get_top_most_different_tuples(dist, n: int):\n",
    "        return list(zip(*np.unravel_index(np.argsort(-dist.ravel()), dist.shape)))[:n]\n",
    "        \n",
    "    def most_different_features(top_number: int) -> None:\n",
    "        for region, x in epigenomes.items():\n",
    "            dist = euclidean_distances(x.T)\n",
    "            most_distance_columns_indices = get_top_most_different(dist, top_number)\n",
    "            columns = x.columns[most_distance_columns_indices]\n",
    "            fig, axes = plt.subplots(nrows=1, ncols=5, figsize=(25, 5))\n",
    "            print(f\"Top {top_number} different features from {region}.\")\n",
    "            for column, axis in zip(columns, axes.flatten()):\n",
    "                head, tail = x[column].quantile([0.05, 0.95]).values.ravel()\n",
    "\n",
    "                mask = ((x[column] < tail) & (x[column] > head)).values\n",
    "\n",
    "                cleared_x = x[column][mask]\n",
    "                cleared_y = labels[region].values.ravel()[mask]\n",
    "\n",
    "                cleared_x[cleared_y==0].hist(ax=axis, bins=20)\n",
    "                cleared_x[cleared_y==1].hist(ax=axis, bins=20)\n",
    "\n",
    "                axis.set_title(column)\n",
    "            fig.tight_layout()\n",
    "            plt.show()\n",
    "            \n",
    "    def most_different_tuples(top_number: int) -> None:\n",
    "        for region, x in epigenomes.items():\n",
    "            dist = euclidean_distances(x.T)\n",
    "            dist = np.triu(dist)\n",
    "            tuples = get_top_most_different_tuples(dist, top_number)\n",
    "            fig, axes = plt.subplots(nrows=1, ncols=5, figsize=(25, 5))\n",
    "            print(f\"Top {top_number} different tuples of features from {region}.\")\n",
    "            for (i, j), axis in zip(tuples, axes.flatten()):\n",
    "                column_i = x.columns[i]\n",
    "                column_j = x.columns[j]\n",
    "                for column in (column_i, column_j):\n",
    "                    head, tail = x[column].quantile([0.05, 0.95]).values.ravel()\n",
    "                    mask = ((x[column] < tail) & (x[column] > head)).values\n",
    "                    x[column][mask].hist(ax=axis, bins=20, alpha=0.5)\n",
    "                axis.set_title(f\"{column_i} and {column_j}\")\n",
    "            fig.tight_layout()\n",
    "            plt.show()\n",
    "        \n",
    "    correlated(3)\n",
    "    uncorrelated(3)\n",
    "    most_different_features(5)\n",
    "    most_different_tuples(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data analysis (selection and decomposition)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_filtered_with_boruta(epigenomes: Dict[str, pd.DataFrame], labels: Dict[str, pd.DataFrame]) -> Dict[str, pd.DataFrame]:\n",
    "    def get_features_filter(data: pd.DataFrame, label: pd.DataFrame, name: str) -> BorutaPy:\n",
    "        boruta_selector = BorutaPy(\n",
    "            RandomForestClassifier(n_jobs=cpu_count(), class_weight='balanced', max_depth=5),\n",
    "            n_estimators='auto',\n",
    "            verbose=2,\n",
    "            alpha=0.05, # p_value\n",
    "            max_iter=10, # In practice one would run at least 100-200 times\n",
    "            random_state=42\n",
    "        )\n",
    "        boruta_selector.fit(data.values, label.values.ravel())\n",
    "        return boruta_selector\n",
    "\n",
    "    return {\n",
    "        region: get_features_filter(\n",
    "            data=data,\n",
    "            label=labels[region],\n",
    "            name=f\"{cell_line}/{region}\"\n",
    "        ).transform(data.values)\n",
    "        for region, data in tqdm(\n",
    "            epigenomes.items(),\n",
    "            desc=\"Running Baruta Feature estimation\"\n",
    "        )\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tasks(epigenomes: Dict[str, pd.DataFrame]):\n",
    "    def check_tasks(xs, ys, titles: List[str]) -> Tuple:\n",
    "        assert len(xs) == len(ys) == len(titles)\n",
    "        for x, y in zip(xs, ys):\n",
    "            assert x.shape[0] == y.shape[0]\n",
    "        return xs, ys, titles\n",
    "    \n",
    "    tasks = {\n",
    "        \"x\":[\n",
    "            *[\n",
    "                val.values\n",
    "                for val in epigenomes.values()\n",
    "            ],\n",
    "            *[\n",
    "                val.values\n",
    "                for val in sequences.values()\n",
    "            ],\n",
    "            pd.concat(sequences.values()).values,\n",
    "            pd.concat(sequences.values()).values,\n",
    "            *[\n",
    "                np.hstack([\n",
    "                    pca(epigenomes[region], n_components=25),\n",
    "                    mfa(sequences[region], n_components=25)\n",
    "                ])\n",
    "                for region in epigenomes\n",
    "            ]\n",
    "        ],\n",
    "        \"y\":[\n",
    "            *[\n",
    "                val.values.ravel()\n",
    "                for val in labels.values()\n",
    "            ],\n",
    "            *[\n",
    "                val.values.ravel()\n",
    "                for val in labels.values()\n",
    "            ],\n",
    "            pd.concat(labels.values()).values.ravel(),\n",
    "            np.vstack([np.ones_like(labels[\"promoters\"]), np.zeros_like(labels[\"enhancers\"])]).ravel(),\n",
    "            *[\n",
    "                val.values.ravel()\n",
    "                for val in labels.values()\n",
    "            ],\n",
    "        ],\n",
    "        \"titles\":[\n",
    "            \"Epigenomes promoters\",\n",
    "            \"Epigenomes enhancers\",\n",
    "            \"Sequences promoters\",\n",
    "            \"Sequences enhancers\",\n",
    "            \"Sequences active regions\",\n",
    "            \"Sequences regions types\",\n",
    "            \"Combined promoters data\",\n",
    "            \"Combined enhancers data\"\n",
    "        ]\n",
    "    }\n",
    "    \n",
    "    return check_tasks(tasks['x'], tasks['y'], tasks['title'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_decomposed_data(xs, ys, titles):\n",
    "    colors = np.array([\n",
    "        \"tab:blue\",\n",
    "        \"tab:orange\",\n",
    "    ])\n",
    "    \n",
    "    def pca(data: np.ndarray, n_components: int = 2) -> np.ndarray:\n",
    "        return PCA(n_components=n_components, random_state=42).fit_transform(data)\n",
    "\n",
    "    def mfa(data: pd.DataFrame, n_components : int = 2, nucleotides : str = 'actg') -> np.ndarray:\n",
    "        return MFA(groups={\n",
    "            nucleotide: [\n",
    "                column\n",
    "                for column in data.columns\n",
    "                if nucleotide in column\n",
    "            ]\n",
    "            for nucleotide in nucleotides\n",
    "        }, n_components=n_components, random_state=42).fit_transform(data)\n",
    "    \n",
    "    def sklearn_tsne(x: np.ndarray, perplexity: int, dimensionality_threshold: int = 50):\n",
    "        if x.shape[1] > dimensionality_threshold:\n",
    "            x = pca(x, n_components=dimensionality_threshold)\n",
    "        return STSNE(perplexity=perplexity, n_jobs=cpu_count(), random_state=42).fit_transform(x)\n",
    "    \n",
    "    def ulyanov_tsne(x: np.ndarray, perplexity: int, dimensionality_threshold: int = 50, n_components: int = 2):\n",
    "        if x.shape[1] > dimensionality_threshold:\n",
    "            x = pca(x, n_components=dimensionality_threshold)\n",
    "        return UTSNE(n_components=n_components, perplexity=perplexity, n_jobs=cpu_count(), random_state=42, verbose=True).fit_transform(x)\n",
    "    \n",
    "    def cannylab_tsne(x: np.ndarray, perplexity: int, dimensionality_threshold: int = 50):\n",
    "        if x.shape[1] > dimensionality_threshold:\n",
    "            x = pca(x, n_components=dimensionality_threshold)\n",
    "        return CTSNE(perplexity=perplexity, random_seed=42).fit_transform(x)\n",
    "    \n",
    "    def nystroem(x:np.array)->np.array:\n",
    "        return Nystroem(random_state=42, n_components=300).fit_transform(x)\n",
    "\n",
    "    def monte_carlo(x:np.array)->np.array:\n",
    "        return RBFSampler(random_state=42, n_components=300).fit_transform(x)\n",
    "\n",
    "    def linear(x:np.array)->np.array:\n",
    "        return x\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_decomposed_data():\n",
    "    def show_pca():\n",
    "        fig, axes = plt.subplots(nrows=2, ncols=4, figsize=(32, 16))\n",
    "        for x, y, title, axis in tqdm(zip(xs, ys, titles, axes.flatten()), desc=\"Computing PCAs\", total=len(xs)):\n",
    "            axis.scatter(*pca(x).T, s=1, color=colors[y])\n",
    "            axis.xaxis.set_visible(False)\n",
    "            axis.yaxis.set_visible(False)\n",
    "            axis.set_title(f\"PCA decomposition - {title}\")\n",
    "        plt.show()\n",
    "        \n",
    "    def show_tsne():\n",
    "        for perpexity in tqdm((30, 40, 50, 100, 500, 5000), desc=\"Running perplexities\"):\n",
    "            fig, axes = plt.subplots(nrows=2, ncols=4, figsize=(40, 20))\n",
    "            for x, y, title, axis in tqdm(zip(xs, ys, titles, axes.flatten()), desc=\"Computing TSNEs\", total=len(xs)):\n",
    "                axis.scatter(*cannylab_tsne(x, perplexity=perpexity).T, s=1, color=colors[y])\n",
    "                axis.xaxis.set_visible(False)\n",
    "                axis.yaxis.set_visible(False)\n",
    "                axis.set_title(f\"TSNE decomposition - {title}\")\n",
    "            fig.tight_layout()\n",
    "            plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
